File: tests/test_xbee_compatibility.py
#!/usr/bin/env python3
"""
Test script to verify XBee compatibility with 0 values
"""

import struct
import sys

def create_xbee_packet(counter):
    """Create a 78-byte packet with all 0 values for XBee compatibility"""
    packet = bytearray()
    
    # Header: FF FF 54 52
    packet.extend([0xFF, 0xFF, 0x54, 0x52])
    
    # Team ID and packet counter
    packet.append(142)  # Team ID
    packet.append(counter % 256)
    
    # 17 float values (68 bytes) - all set to 0 for XBee
    float_values = [
        0.0,  # altitude_agl
        0.0,  # rocket_gps_altitude
        0.0,  # rocket_latitude
        0.0,  # rocket_longitude
        0.0,  # payload_gps_altitude
        0.0,  # payload_latitude
        0.0,  # payload_longitude
        0.0,  # stage_gps_altitude
        0.0,  # stage_latitude
        0.0,  # stage_longitude
        0.0,  # gyro_x
        0.0,  # gyro_y
        0.0,  # gyro_z
        0.0,  # accel_x
        0.0,  # accel_y
        0.0,  # accel_z
        0.0   # angle
    ]
    
    for value in float_values:
        packet.extend(struct.pack('<f', value))
    
    # Status code (0 for XBee)
    packet.append(0)
    
    # Calculate checksum (sum of bytes 4-75, mod 256)
    checksum = sum(packet[4:75]) % 256
    packet.append(checksum)
    
    # Footer: 0D 0A
    packet.extend([0x0D, 0x0A])
    
    return packet

def test_xbee_packet():
    """Test XBee packet format"""
    print("🧪 Testing XBee Packet Format")
    print("=" * 40)
    
    packet = create_xbee_packet(1)
    
    print(f"📦 Packet length: {len(packet)} bytes")
    print(f"🔍 Header: {[f'{b:02X}' for b in packet[:4]]}")
    print(f"🏷️  Team ID: {packet[4]}")
    print(f"🔢 Counter: {packet[5]}")
    print(f"📊 Status: {packet[74]}")
    print(f"🔢 Checksum: {packet[75]:02X}")
    print(f"📄 Footer: {[f'{b:02X}' for b in packet[-2:]]}")
    
    # Verify all float values are 0
    float_start = 6
    float_count = 17
    all_zeros = True
    
    for i in range(float_count):
        offset = float_start + (i * 4)
        if offset + 4 <= len(packet):
            value = struct.unpack('<f', packet[offset:offset+4])[0]
            if value != 0.0:
                all_zeros = False
                print(f"❌ Float {i}: {value} (should be 0)")
    
    if all_zeros:
        print("✅ All float values are 0 (XBee compatible)")
    else:
        print("❌ Some float values are not 0")
    
    # Verify checksum
    calculated_checksum = sum(packet[4:75]) % 256
    if calculated_checksum == packet[75]:
        print("✅ Checksum is correct")
    else:
        print(f"❌ Checksum mismatch: calculated={calculated_checksum:02X}, stored={packet[75]:02X}")
    
    print(f"\n🎯 XBee compatibility test completed!")
    return all_zeros and calculated_checksum == packet[75]

if __name__ == "__main__":
    success = test_xbee_packet()
    sys.exit(0 if success else 1) 

File: tests/test_xbee_compatibility.js
#!/usr/bin/env node
/**
 * test_xbee_compatibility.js
 * Verifies XBee packet format with all-zero float values.
 * JS port of Tests/test_xbee_compatibility.py
 */

'use strict';

function createXbeePacket(counter) {
  const packet = Buffer.alloc(78, 0);
  let offset = 0;

  // Header FF FF 54 52
  packet[offset++] = 0xff;
  packet[offset++] = 0xff;
  packet[offset++] = 0x54;
  packet[offset++] = 0x52;

  packet[offset++] = 142;          // Team ID
  packet[offset++] = counter & 0xff;

  // 17 floats all = 0.0
  for (let i = 0; i < 17; i++) {
    packet.writeFloatLE(0.0, offset);
    offset += 4;
  }

  // Status = 0
  packet[offset++] = 0;

  // Checksum
  let sum = 0;
  for (let i = 4; i < 75; i++) sum += packet[i];
  packet[offset++] = sum & 0xff;

  // Footer
  packet[offset++] = 0x0d;
  packet[offset]   = 0x0a;

  return packet;
}

function testXbeePacket() {
  console.log('🧪 Testing XBee Packet Format');
  console.log('='.repeat(40));

  const packet = createXbeePacket(1);
  let allPassed = true;

  console.log(`📦 Packet length: ${packet.length} bytes`);
  console.log(`🔍 Header: [${[...packet.slice(0,4)].map(b=>'0x'+b.toString(16).padStart(2,'0')).join(', ')}]`);
  console.log(`🏷️  Team ID: ${packet[4]}`);
  console.log(`🔢 Counter: ${packet[5]}`);
  console.log(`📊 Status: ${packet[74]}`);
  console.log(`🔢 Checksum: 0x${packet[75].toString(16).toUpperCase().padStart(2,'0')}`);
  console.log(`📄 Footer: [0x${packet[76].toString(16).toUpperCase()}, 0x${packet[77].toString(16).toUpperCase()}]`);

  // Verify all floats = 0
  let allZeros = true;
  const floatStart = 6;
  for (let i = 0; i < 17; i++) {
    const val = packet.readFloatLE(floatStart + i * 4);
    if (val !== 0.0) { allZeros = false; console.error(`❌ Float ${i}: ${val} (expected 0)`); }
  }
  if (allZeros) console.log('✅ All float values are 0 (XBee compatible)');
  else { allPassed = false; }

  // Verify checksum
  let calcSum = 0;
  for (let i = 4; i < 75; i++) calcSum += packet[i];
  calcSum &= 0xff;
  if (calcSum === packet[75]) console.log('✅ Checksum is correct');
  else {
    console.error(`❌ Checksum mismatch: calc=0x${calcSum.toString(16).toUpperCase()}, stored=0x${packet[75].toString(16).toUpperCase()}`);
    allPassed = false;
  }

  console.log(`\n🎯 XBee compatibility test ${allPassed ? 'PASSED' : 'FAILED'}!\n`);
  return allPassed && allZeros;
}

const ok = testXbeePacket();
process.exit(ok ? 0 : 1);